nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/PutHDFS.java

```diff
@@ -372,54 +370,56 @@ public Object run() {
                     final StopWatch stopWatch = new StopWatch(true);
-                    session.read(putFlowFile, new InputStreamCallback() {
-
-                        @Override
-                        public void process(InputStream in) throws IOException {
-                            OutputStream fos = null;
-                            Path createdFile = null;
-                            try {
-                                if (conflictResponse.equals(APPEND_RESOLUTION) && destinationExists) {
-                                    fos = hdfs.append(copyFile, bufferSize);
-                                } else {
-                                    final EnumSet<CreateFlag> cflags = EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE);
-
-                                    if (shouldIgnoreLocality(context, session)) {
-                                        cflags.add(CreateFlag.IGNORE_CLIENT_LOCALITY);
-                                    }
+                    session.read(putFlowFile, in -> {
+                        OutputStream fos = null;
+                        Path createdFile = null;
+                        try {
+                            if (conflictResponse.equals(APPEND_RESOLUTION) && destinationExists) {
+                                fos = hdfs.append(copyFile, bufferSize);
+                            } else {
+                                final EnumSet<CreateFlag> cflags = EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE);
 
-                                    fos = hdfs.create(actualCopyFile, FsCreateModes.applyUMask(FsPermission.getFileDefault(),
-                                            FsPermission.getUMask(hdfs.getConf())), cflags, bufferSize, replication, blockSize,
-                                            null, null);
+                                if (shouldIgnoreLocality(context, session)) {
+                                    cflags.add(CreateFlag.IGNORE_CLIENT_LOCALITY);
                                 }
 
-                                if (codec != null) {
-                                    fos = codec.createOutputStream(fos);
+                                fos = hdfs.create(actualCopyFile, FsCreateModes.applyUMask(FsPermission.getFileDefault(),
+                                        FsPermission.getUMask(hdfs.getConf())), cflags, bufferSize, replication, blockSize,
+                                        null, null);
+                            }
+
+                            if (codec != null) {
+                                fos = codec.createOutputStream(fos);
+                            }
+                            createdFile = actualCopyFile;
+                            BufferedInputStream bis = new BufferedInputStream(in);
+                            StreamUtils.copy(bis, fos);
+                            bis = null;
+                            fos.flush();
+                        } catch (IOException e) {
```

Comment:
There might be ambiguity regarding where the exception should be handled. If `hdfs.create` threw an exception shouldn't we get out of this catch block and skip the lines from 424 to 465? Even if we successfully reset the HDFS resources, in the current run the processor fails to create files on the HDFS. To preserve the original behavior a custom RuntimeException could be thrown and caught by the outer catch block. Since we are in a session.write, we cannot throw the GSSException wrapped in an IOException because it will be caught by the framework before the outer catch block.

Reply:
Works for me, the key was to not have it blow up for a "simple" relogin error. I thought I'd found that depending on the config the first create() could cause the login error. If anything but a "Kerberos relogin error" occurs, I think we should rightfully propagate the exception, I can throw it as a ProcessException inside the read(). What do you think?

Reply:
Yeah I think it's fine. If executing the mentioned lines does not cause discrepancy, I prefer the simplicity.